File: native/android/app/src/main/java/com/mediacleaner/nativeapp/data/UpdateClient.kt
package com.mediacleaner.nativeapp.data

import com.mediacleaner.nativeapp.BuildConfig
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URI
import java.net.URL
import java.io.ByteArrayOutputStream
import android.os.SystemClock

data class NativeUpdate(
    val version: String,
    val build: Long,
    val preview: Boolean,
    val downloadUrl: String,
    val notes: String,
) {
    val label get() = "$version ($build)"
    fun download(source: Int): String = when (source) {
        1 -> "https://ghproxy.net/$downloadUrl"
        2 -> "https://gh-proxy.com/$downloadUrl"
        else -> downloadUrl
    }

    fun compare(other: NativeUpdate): Int {
        val ours = version.split('.').map { it.toLong() }
        val theirs = other.version.split('.').map { it.toLong() }
        for (index in 0..2) if (ours[index] != theirs[index]) return ours[index].compareTo(theirs[index])
        if (preview != other.preview) return if (preview) -1 else 1
        return build.compareTo(other.build)
    }
}

class UpdateClient {
    private val repository = "youmikk/media-cleaner"
    private val tagPattern = Regex("^native-v([0-9]+\\.[0-9]+\\.[0-9]+)-(stable|preview)\\.([0-9]+)$")
    private val manifestPath = "$repository/main/native/releases.json"
    private val endpoints = listOf(
        "https://cdn.jsdelivr.net/gh/$repository@main/native/releases.json",
        "https://raw.githubusercontent.com/$manifestPath",
        "https://gh-proxy.com/https://raw.githubusercontent.com/$manifestPath",
        "https://api.github.com/repos/$repository/releases?per_page=100",
        "https://gh-proxy.com/https://api.github.com/repos/$repository/releases?per_page=100",
    )

    suspend fun check(includePreview: Boolean): NativeUpdate? = withContext(Dispatchers.IO) {
        val current = NativeUpdate(BuildConfig.VERSION_NAME, BuildConfig.VERSION_CODE.toLong(), BuildConfig.UPDATE_PRERELEASE, "", "")
        var reachedSource = false
        for (endpoint in endpoints) {
            ensureActive()
            try {
                val body = request(endpoint)
                val candidates = if (endpoint.contains("/releases?")) releases(JSONArray(body)) else manifest(JSONObject(body))
                reachedSource = true
                val update = candidates.filter { includePreview || !it.preview }
                    .filter { it.build >= current.build && it.compare(current) > 0 }.maxWithOrNull { a, b -> a.compare(b) }
                if (update != null) return@withContext update
            } catch (e: CancellationException) { throw e }
            catch (e: Exception) { /* A blocked or invalid source falls through to the next mirror. */ }
        }
        if (reachedSource) return@withContext null
        throw IllegalStateException("Update sources unavailable")
    }

    private fun request(address: String): String {
        val connection = URL(address).openConnection() as HttpURLConnection
        try {
            connection.connectTimeout = 5000; connection.readTimeout = 5000
            connection.setRequestProperty("Accept", "application/json")
            connection.setRequestProperty("User-Agent", "MediaCleaner/${BuildConfig.VERSION_NAME}")
            connection.setRequestProperty("Cache-Control", "no-cache")
            check(connection.responseCode == 200)
            val deadline = SystemClock.elapsedRealtime() + 8000
            return connection.inputStream.use { input ->
                val output = ByteArrayOutputStream()
                val buffer = ByteArray(8192)
                while (true) {
                    check(SystemClock.elapsedRealtime() < deadline)
                    val count = input.read(buffer)
                    if (count < 0) break
                    check(output.size() + count <= 1024 * 1024)
                    output.write(buffer, 0, count)
                }
                output.toString("UTF-8")
            }
        } finally { connection.disconnect() }
    }

    private fun manifest(value: JSONObject): List<NativeUpdate> {
        require(value.getInt("schemaVersion") == 1 && value.getString("applicationId") == BuildConfig.APPLICATION_ID)
        val channels = value.getJSONObject("channels")
        return listOf("stable", "preview").mapNotNull { channel ->
            val value = channels.getJSONObject(channel).optJSONObject("android") ?: return@mapNotNull null
            entry(value.getString("tag"), channel == "preview", value.getString("downloadUrl"), value.optString("notes"))
                ?: throw IllegalArgumentException("Invalid update entry")
        }
    }

    private fun releases(values: JSONArray): List<NativeUpdate> = buildList {
        for (index in 0 until values.length()) {
            val release = values.getJSONObject(index)
            if (release.optBoolean("draft")) continue
            val assets = release.optJSONArray("assets") ?: continue
            for (assetIndex in 0 until assets.length()) {
                val asset = assets.getJSONObject(assetIndex)
                if (asset.optString("name") != "MediaCleaner-Native-Android.apk") continue
                entry(release.optString("tag_name"), release.optBoolean("prerelease"),
                    asset.optString("browser_download_url"), release.optString("body"))?.let { add(it) }
            }
        }
    }

    private fun entry(tag: String, preview: Boolean, address: String, notes: String): NativeUpdate? {
        val match = tagPattern.matchEntire(tag) ?: return null
        if ((match.groupValues[2] == "preview") != preview) return null
        val uri = URI(address)
        if (uri.scheme != "https" || uri.host != "github.com" || uri.userInfo != null || uri.port != -1 ||
            uri.query != null || uri.fragment != null || uri.path != "/$repository/releases/download/$tag/MediaCleaner-Native-Android.apk") return null
        val build = match.groupValues[3].toLongOrNull()?.takeIf { it > 0 } ?: return null
        if (match.groupValues[1].split('.').any { it.toLongOrNull() == null }) return null
        return NativeUpdate(match.groupValues[1], build, preview, address, notes.take(16000))
    }
}

File: native/ios/MediaCleanerNative/Services/UpdateClient.swift
import Foundation

struct NativeUpdate: Identifiable {
    let version: String
    let build: Int
    let preview: Bool
    let downloadURL: URL
    let notes: String
    var id: String { "\(version)-\(preview)-\(build)" }
    var label: String { "\(version) (\(build))" }

    func download(source: Int) -> URL {
        let prefix = source == 1 ? "https://ghproxy.net/" : source == 2 ? "https://gh-proxy.com/" : ""
        return URL(string: prefix + downloadURL.absoluteString)!
    }
    func newer(than other: NativeUpdate) -> Bool {
        let ours = version.split(separator: ".").map { Int($0)! }
        let theirs = other.version.split(separator: ".").map { Int($0)! }
        for index in 0..<3 where ours[index] != theirs[index] { return ours[index] > theirs[index] }
        if preview != other.preview { return !preview }
        return build > other.build
    }
}

struct UpdateClient {
    static let repository = "youmikk/media-cleaner"
    static let currentVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "0.0.1"
    static let currentBuild = Int(Bundle.main.infoDictionary?["CFBundleVersion"] as? String ?? "1") ?? 1
    static let currentPreview = (Bundle.main.infoDictionary?["MCPrerelease"] as? String ?? "YES") == "YES"
    private let decoder = JSONDecoder()

    private struct Feed: Decodable {
        let schemaVersion: Int
        let applicationId: String
        let channels: [String: [String: Entry]]
    }
    private struct Entry: Decodable { let tag: String; let downloadUrl: String; let notes: String }
    private struct Release: Decodable {
        let tag_name: String
        let prerelease: Bool
        let draft: Bool
        let body: String?
        let assets: [Asset]
    }
    private struct Asset: Decodable { let name: String; let browser_download_url: String }

    func check(includePreview: Bool) async throws -> NativeUpdate? {
        let repo = Self.repository
        let endpoints = [
            "https://cdn.jsdelivr.net/gh/\(repo)@main/native/releases.json",
            "https://raw.githubusercontent.com/\(repo)/main/native/releases.json",
            "https://gh-proxy.com/https://raw.githubusercontent.com/\(repo)/main/native/releases.json",
            "https://api.github.com/repos/\(repo)/releases?per_page=100",
            "https://gh-proxy.com/https://api.github.com/repos/\(repo)/releases?per_page=100",
        ]
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 5; config.timeoutIntervalForResource = 8
        let session = URLSession(configuration: config)
        defer { session.invalidateAndCancel() }
        let current = NativeUpdate(version: Self.currentVersion, build: Self.currentBuild, preview: Self.currentPreview,
            downloadURL: URL(string: "https://github.com/\(repo)")!, notes: "")
        var reachedSource = false
        for endpoint in endpoints {
            try Task.checkCancellation()
            do {
                var request = URLRequest(url: URL(string: endpoint)!, cachePolicy: .reloadIgnoringLocalCacheData)
                request.setValue("application/json", forHTTPHeaderField: "Accept")
                request.setValue("MediaCleaner/\(Self.currentVersion)", forHTTPHeaderField: "User-Agent")
                let (data, response) = try await session.data(for: request)
                guard (response as? HTTPURLResponse)?.statusCode == 200, data.count <= 1024 * 1024 else { throw URLError(.badServerResponse) }
                let candidates = try (endpoint.contains("/releases?") ? releases(data) : manifest(data))
                reachedSource = true
                let update = candidates.filter { (includePreview || !$0.preview) && $0.newer(than: current) }
                    .max { $1.newer(than: $0) }
                if let update { return update }
            } catch {
                if Task.isCancelled { throw CancellationError() }
                // Blocked hosts and invalid responses fall through to the next source.
            }
        }
        if reachedSource { return nil }
        throw URLError(.cannotConnectToHost)
    }

    private func manifest(_ data: Data) throws -> [NativeUpdate] {
        let feed = try decoder.decode(Feed.self, from: data)
        guard feed.schemaVersion == 1, feed.applicationId == Bundle.main.bundleIdentifier,
              feed.channels["stable"] != nil, feed.channels["preview"] != nil else { throw URLError(.cannotParseResponse) }
        return try ["stable", "preview"].compactMap { channel in
            guard let value = feed.channels[channel]?["ios"] else { return nil }
            guard let result = entry(tag: value.tag, preview: channel == "preview", address: value.downloadUrl, notes: value.notes) else {
                throw URLError(.cannotParseResponse)
            }
            return result
        }
    }
    private func releases(_ data: Data) throws -> [NativeUpdate] {
        try decoder.decode([Release].self, from: data).compactMap { release in
            guard !release.draft, let asset = release.assets.first(where: { $0.name == "MediaCleaner-Native-iOS-unsigned.ipa" }) else { return nil }
            return entry(tag: release.tag_name, preview: release.prerelease, address: asset.browser_download_url, notes: release.body ?? "")
        }
    }
    private func entry(tag: String, preview: Bool, address: String, notes: String) -> NativeUpdate? {
        let pattern = try! NSRegularExpression(pattern: #"^native-v([0-9]+\.[0-9]+\.[0-9]+)-(stable|preview)\.([0-9]+)$"#)
        let source = tag as NSString
        guard let match = pattern.firstMatch(in: tag, range: NSRange(location: 0, length: source.length)) else { return nil }
        let version = source.substring(with: match.range(at: 1))
        guard (source.substring(with: match.range(at: 2)) == "preview") == preview,
              let build = Int(source.substring(with: match.range(at: 3))), build > 0,
              version.split(separator: ".").allSatisfy({ Int($0) != nil }),
              let url = URL(string: address), url.scheme == "https", url.host == "github.com",
              url.user == nil, url.password == nil, url.port == nil, url.query == nil, url.fragment == nil,
              url.path == "/\(Self.repository)/releases/download/\(tag)/MediaCleaner-Native-iOS-unsigned.ipa" else { return nil }
        return NativeUpdate(version: version, build: build, preview: preview, downloadURL: url, notes: String(notes.prefix(16000)))
    }
}
